### PR TITLE
feat: add character translation to data extract tool

### PR DESCRIPTION
**Pull request by jwilliams-ocient**
# Overview
Extract config allows users to apply a UTF-8 character translation to all VARCHAR data extracted.

There are 3 new configuration properties:

1. `translate_characters_mode` which is either "CHAR" or "HEX"
2. `translate_characters_from`
   - a flat UTF-8 string when `translate_characters_mode == "CHAR"`
   - a list of hex strings when `translate_characters_mode == "HEX"`
3. `translate_characters_to`
   - a flat UTF-8 string when `translate_characters_mode == "CHAR"`
   - a list of hex strings when `translate_characters_mode == "HEX"`

The elements of the "from" and "to" lists define the translation between source and destination character sets.

Please refer to [confluence](https://ocient.atlassian.net/wiki/spaces/PM/pages/3393388586/Requirements+-+Extract+To+Translation+Feature) for the complete list of feature requirements.

# Checklist
- [x] Add Unit Tests
- [x] Add Release Note
- [x] Add user doc (xgsrc repo)

### CODE REVIEW

File: src/main/java/com/ocient/cli/extract/ExtractConfiguration.java [hunk 1]
```diff
@@ -87,6 +104,12 @@ public ExtractConfiguration(final Properties properties)
         BASE64,
     }
 
+    public enum TranslateCharactersMode
+    {
+        HEXADECIMAL,
```
Comment (jwilliams-ocient) [inline src/main/java/com/ocient/cli/extract/ExtractConfiguration.java:109]: rename this to "HEX" it's annoying that you have to type of "HEXADECIMAL"

File: src/main/java/com/ocient/cli/extract/ExtractConfiguration.java [hunk 2]
```diff
@@ -351,4 +393,114 @@ private void validateConfiguration() throws ParseException
...
+                    return new TranslateCharactersConfig(
+                        Streams
+                            .zip(
+                                Arrays.stream(fromHexArray).map(TranslateCharactersConfig::hexToCodePoint),
+                                Arrays.stream(toHexArray).map(TranslateCharactersConfig::hexToCodePoint), 
+                                Map::entry)
+                            .collect(Collectors.toUnmodifiableMap(Entry::getKey, Entry::getValue, (t, u) -> u)));
+            }
+    
+            throw new ParseException(String.format("Unsupported %s", TRANSLATE_CHARACTERS_MODE));
```
Comment (jwilliams-ocient) [inline src/main/java/com/ocient/cli/extract/ExtractConfiguration.java:477]: this is incomplete

**Comment by daveboutcher** [comment]
I am missing something here. Where is the part that you decide that  c̦  ȼ̓ and č̣ should map tp "c"

File: src/main/java/com/ocient/cli/extract/ExtractConfiguration.java [hunk 3]
```diff
@@ -351,4 +393,114 @@ private void validateConfiguration() throws ParseException
...
+                     * Input strings are interpreted as a UTF-8 string where each character in `fromChars`
+                     * maps to the character at the same index in `toChars`
+                     * 
+                     * For example,
+                     * 
+                     * translate_characters_mode="char",
+                     * translate_characters_from="pz"
+                     * translate_characters_to  ="lo"
+                     * 
+                     * e.g. "Heppz" => "Hello"
```
Comment (jwilliams-ocient) [inline src/main/java/com/ocient/cli/extract/ExtractConfiguration.java:440]: @daveboutcher a couple of unit tests probably would have helped here 🙂 . 

- In "CHAR" mode, the mapping is defined using the indices in the "from"/"to" strings `from.charAt(n) => to.charAt(n)`. In the example above, the translations defined are `p => l` & `z => o`.
- In "HEX" mode, the mapping is defined using the indices in the "from"/"to" arrays `from[n] => to[n]`

**Comment by jkelley-Ocient** [comment]
Working on user doc and will PR this into XGSRC shortly

**Pull request merged**